lib/src/chunking.rs

```diff
@@ -372,82 +394,409 @@ impl Chunking {
     let mut r = Vec::new();
-    // And handle the easy case of enough bins for all components
-    // TODO: Possibly try to split off large files?
-    if components.len() <= bins.get() as usize {
-        r.extend(components.iter().map(|v| vec![v]));
-        return r;
-    }
-    // Create a mutable copy
     let mut components: Vec<_> = components.iter().collect();
-    // Iterate over the component tail, folding by source id
-    let mut by_src = HashMap::<_, Vec<&ObjectSourceMetaSized>>::new();
-    // Take the tail off components, then build up mapping from srcid -> Vec<component>
-    for component in components.split_off(bins.get() as usize) {
-        by_src
-            .entry(&component.meta.srcid)
-            .or_default()
-            .push(component);
+    let before_processing_pkgs_len = components.len();
+    if before_processing_pkgs_len == 0 {
+        return Vec::new();
     }
-    // Take all the non-tail (largest) components, and append them first
-    r.extend(components.into_iter().map(|v| vec![v]));
-    // Add the tail
-    r.extend(by_src.into_values());
-    // And order the new list
-    sort_packing(&mut r);
-    // It's possible that merging components gave us enough space; if so
-    // we're done!
-    if r.len() <= bins.get() as usize {
-        return r;
+    //Flatten out prior_build_metadata[i] to view all the packages in prior build as a single vec
+    //
+    //If the current rpm-ostree commit to be encapsulated is not the one in which packing structure changes, then
+    //  Compare flatten(prior_build_metadata[i]) to components to see if pkgs added, updated,
+    //  removed or kept same
+    //  if pkgs added, then add them to the last bin of prior[i][n]
+    //  if pkgs removed, then remove them from the prior[i]
+    //  iterate through prior[i] and make bins according to the name in nevra of pkgs and return
+    //  (no need of recomputing packaging structure)
+    //else if pkg structure to be changed || prior build not specified
+    //  Recompute optimal packaging strcuture (Compute partitions, place packages and optimize build)
+
+    if let Some(prior_build) = prior_build_metadata
+    /* && structure not be changed*/
+    {
+        println!("Keeping old package structure");
+        let mut curr_build: Vec<Vec<String>> = prior_build.clone();
+        //Packing only manaages RPMs not OStree commit
+        curr_build.remove(0);
+        let mut prev_pkgs: Vec<String> = Vec::new();
+        for bin in &curr_build {
+            for pkg in bin {
+                prev_pkgs.push(pkg.to_string());
+            }
+        }
+        prev_pkgs.retain(|name| !name.is_empty());
+        let curr_pkgs: Vec<String> = components
+            .iter()
+            .map(|pkg| pkg.meta.name.to_string())
+            .collect();
+        let prev_pkgs_set: HashSet<String> = HashSet::from_iter(prev_pkgs);
+        let curr_pkgs_set: HashSet<String> = HashSet::from_iter(curr_pkgs);
```

Comment:
Then we're taking the vector copy and making a `HashSet`?  I think we can just directly do:

```suggestion
        let curr_pkgs_set: HashSet<_> = components.iter().map(|pkg| pkg.meta.name.as_str()).collect();
```

And this will make a set of *borrowed* strings instead of a deep copy.

Reply:
This would lead to the prev_pkgs_set be Vec<String> and the curr_pkgs_set to be Vec<Rc<str>> leading to incosistency

Reply:
Aren't in both cases we're borrowing a `&str`?

Reply:
OK this compiles at least, WDYT?

```
diff --git a/lib/src/chunking.rs b/lib/src/chunking.rs
index 18229a51..f8af0241 100644
--- a/lib/src/chunking.rs
+++ b/lib/src/chunking.rs
@@ -571,21 +571,14 @@ fn basic_packing<'a>(
         let mut curr_build: Vec<Vec<String>> = prior_build.clone();
         //Packing only manaages RPMs not OStree commit
         curr_build.remove(0);
-        let mut prev_pkgs: Vec<String> = Vec::new();
-        for bin in &curr_build {
-            for pkg in bin {
-                prev_pkgs.push(pkg.to_string());
-            }
-        }
-        prev_pkgs.retain(|name| !name.is_empty());
-        let curr_pkgs: Vec<String> = components
+        let prev_pkgs_set: HashSet<_> = curr_build
             .iter()
-            .map(|pkg| pkg.meta.name.to_string())
+            .flat_map(|b| b.iter().map(|p| p.as_str()))
+            .filter(|&name| name != "")
             .collect();
-        let prev_pkgs_set: HashSet<String> = HashSet::from_iter(prev_pkgs);
-        let curr_pkgs_set: HashSet<String> = HashSet::from_iter(curr_pkgs);
-        let added: HashSet<&String> = curr_pkgs_set.difference(&prev_pkgs_set).collect();
-        let removed: HashSet<&String> = prev_pkgs_set.difference(&curr_pkgs_set).collect();
+        let curr_pkgs_set: HashSet<_> = components.iter().map(|pkg| &*pkg.meta.name).collect();
+        let added: HashSet<_> = curr_pkgs_set.difference(&prev_pkgs_set).collect();
+        let removed: HashSet<_> = prev_pkgs_set.difference(&curr_pkgs_set).collect();
         let mut add_pkgs_v: Vec<String> = Vec::new();
         for pkg in added {
             add_pkgs_v.push(pkg.to_string());

```

Reply:
Note e.g. that `&*pkg.meta.name` lets us go from `Rc<str>` to `&str`.

Reply:
ok makes sense

Reply:
Did you want to try taking this patch and integrating it?  I think there's some more cleanup along these lines we could do below this too

Reply:
it makes the `removed` type to `HashSet<&&str>` causing problem here:
```
for bin in curr_build.iter_mut() {
            bin.retain(|pkg| !removed.contains(pkg));
        }
```

Reply:
OK, this is tricky.  I think what we're running up against here is that we're passing around a `Vec<String>` in many places.  A lot of the rest of this code uses `Rc<str>` precisely as a way to share strings in multiple data structures.

I think it'd help to make a type alias.  If we did e.g. 

```
type PackageList = Vec<String>;
```
then it'd be a lot more self describing particularly when used inside a `Vec` itself, i.e. we go from `Vec<Vec<String>>` to `Vec<PackageList>`.

There's another thing here which is that I think in general we don't care about ordering *until* we go to serialize it (in which case it should be sorted).

So we could also make this

```
type PackageSet = HashSet<String>;
```

But, combining all this it'd be

```
type PackageSet = HashSet<Rc<str>>;
```

?

Reply:
This leads to a borrowed reference when defining sets to compute set differences on and causes errors